src/KubeOps/Operator/Rbac/RbacBuilder.cs

```diff
@@ -0,0 +1,72 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using k8s.Models;
+using KubeOps.Operator.Builder;
+
+namespace KubeOps.Operator.Rbac
+{
+    internal class RbacBuilder : IRbacBuilder
+    {
+        private readonly List<Type> _componentTypes;
+
+        private readonly bool _hasWebhooks;
+
+        public RbacBuilder(IComponentRegistrar componentRegistrar)
+        {
+            var controllerTypes = componentRegistrar.ControllerRegistrations
+                .Select(t => t.ControllerType).ToList();
+            var finalizerTypes = componentRegistrar.FinalizerRegistrations
+                .Select(t => t.FinalizerType).ToList();
+            var validatorTypes = componentRegistrar.ValidatorRegistrations
+                .Select(t => t.ValidatorType).ToList();
+            var mutatorTypes = componentRegistrar.MutatorRegistrations
+                .Select(t => t.MutatorType).ToList();
+
+            _componentTypes = Enumerable.Empty<Type>()
+                .Concat(controllerTypes)
+                .Concat(finalizerTypes)
+                .Concat(validatorTypes)
+                .Concat(mutatorTypes)
+                .Distinct()
+                .ToList();
+
+            _hasWebhooks = validatorTypes.Any() || mutatorTypes.Any();
+        }
+
+        public V1ClusterRole BuildManagerRbac()
+        {
+            var entityRbacPolicyRules = GetAttributes<EntityRbacAttribute>()
+                .SelectMany(attribute => attribute.CreateRbacPolicies());
+
+            var genericRbacPolicyRules = GetAttributes<GenericRbacAttribute>()
+                .Select(attribute => attribute.CreateRbacPolicy());
+
+            var rules = entityRbacPolicyRules.Concat(genericRbacPolicyRules).ToList();
+
+            if (_hasWebhooks)
+            {
+                var servicePolicies = new EntityRbacAttribute(
+                    typeof(V1Service),
+                    typeof(V1ValidatingWebhookConfiguration))
+                {
+                    Verbs = RbacVerb.Get | RbacVerb.Create | RbacVerb.Update | RbacVerb.Patch,
+                }.CreateRbacPolicies();
+
+                rules = rules.Concat(servicePolicies).ToList();
+            }
+
+            return new V1ClusterRole(
+                null,
+                $"{V1ClusterRole.KubeGroup}/{V1ClusterRole.KubeApiVersion}",
+                V1ClusterRole.KubeKind,
+                new V1ObjectMeta { Name = "operator-role" },
+                new List<V1PolicyRule>(rules));
+        }
+
+        private IEnumerable<TAttribute> GetAttributes<TAttribute>()
+            where TAttribute : Attribute =>
+            _componentTypes.SelectMany(type => type.GetCustomAttributes<TAttribute>(true));
```

Comment:
If I understand this correctly, the "new" implementation scans for Rbac attributes _only_ on the registered controllers/finalizers/webhooks? Because the old implementation allowed the rbac attributes on any type in the assembly. It kind of makes sense to limit rbac attributes to those entities, but would it therefore be a breaking change? wdyt?

Reply:
Yes, this only scans registered types.

If you look at the `AssemblyScanner` again though, default behavior is still to register all controllers/finalizers/webhooks that are present. This keeps the same behavior as before, scanning entry and executing assemblies automatically and allowing `AddResourceAssembly` to provide more.

I tried to keep all my changes as compatible as possible from an external standpoint.

Reply:
I agree with the argumentation that all types ("relevant" types) are scanned. However, this means a breaking change if we commit this. I'm fine with breaking changes. I just want to clarify the matter to see if it is necessary. Before, one could add an Rbac attribute to the `Program.cs` file because they like to have them there. But I like the clean solution this creates.

Rbac attributes are required to be on a relevant Kubernetes type now.